react/react-hooks/why-are-react-hooks-required.md

```diff
@@ -0,0 +1,232 @@
+
+```javascript
+const ExampleComponent = React.createClass({
+  getInitialState () {
+    return {
+        // ...
+    }
+  },
+  componentDidMount () {
+    // setup
+    // add listener for feature 1
+    // add listener for feature 2
+    // and so on...
+  },
+  componentDidUpdate () {
+    // logic to update state
+  },
+  componentWillUnmount () {
+    // clean up
+    // remove listener for feature 1
+    // remove listener for feature 2
+    // and so on...
+  },
+
+  render() {
+    return (
+      // ...
+    )
+  }
+})
+```
+
+### Post ES6 release
+
+With release of ES6, `class` keyword was introduced in JavaScript and it was a new way to create classes in JavaScript. Later, React deprecated the `createClass` API and moved to `React.Component` API. Now we could create React Components using JavaScript native way of creating classes, as shown in the snippet below.
+
+```js
+class ExampleComponent extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      // ...
+    };
+
+    this.exampleFunction = this.exampleFunction.bind(this);
+  }
+  componentDidMount() {
+    // setup
+    // add listener for feature 1
+    // add listener for feature 2
+    // and so on...
+  }
+
+  componentDidUpdate() {
+    // logic to update state
+  }
+
+  componentWillUnmount () {
+    // clean up
+    // remove listener for feature 1
+    // remove listener for feature 2
+    // and so on...
+  },
+
+  exampleFunction() {
+    //...
+
+    this.setState({
+      // ...
+    });
+  }
+
+  render() {
+    // ...
+  }
+}
+```
+
+With this change, React got better than earlier, but it still had some tradeoffs. Everytime while creating a component, it was compulsory to use constructor and call `super(props)` and also `.bind` every function using `.this` keyword in the constructor.
+
+Now these weren't too big of an issue and also these weren't actually React issues because this is how classes worked in JavaScript. But still, there had to be some more efficient way to overcome this problem.
+
+Later, something called as Class Fields were released, which allowed us to directly add properties to the class, without having to use constructor and we no longer needed to `.bind` any methods in the constructor as we could use arrow functions to create methods in class.
+
+```js
+class Example Component extends React.Component {
+    state = {
+        // ...
+    }
+
+    exampleFunction = () => {
+        // ...
+    }
+}
+```
+
+### React Functional Components
+
+Then with Functional Components, there was a simpler way to create components which did not require any state management or lifecycle methods. Initially, Functional Components were fairly simple. They would take in props and simply render a UI, no other complexity was involved.
```

Comment:
It would be nice to say that functional components were initially stateless components. Then say that they would take prop.... Your choice.